rtmr/extend/extend.go

```diff
@@ -0,0 +1,199 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package extend
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/sha512" // Registrer SHA384 and SHA512
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/google/go-configfs-tsm/configfs/configfsi"
+	"github.com/google/go-configfs-tsm/configfs/linuxtsm"
+	"github.com/google/logger"
+)
+
+const (
+	RtmrSubsystem = "rtmrs"
+	TsmRtmrPrefix = configfsi.TsmPrefix + "/" + RtmrSubsystem
+	TsmRtmrDigest = "digest"
+	TsmPathIndex  = "index"
+	TsmPathTcgMap = "tcg_map"
+)
+
+// RtmrExtend is a struct that represents a rtmr entry in the configfs.
+type RtmrExtend struct {
+	RtmrIndex int
+	entry     *configfsi.TsmPath
+	client    configfsi.Client
+}
+
+// Remove the null bytes from the c-string
+func convertCstring(b []byte) string {
+	return string(bytes.Trim(b, "\x00"))
+}
+
+func (r *RtmrExtend) attribute(subtree string) string {
+	a := *r.entry
+	a.Attribute = subtree
+	return a.String()
+}
+
+// ExtendDigest extends the measurement to the rtmr with the given hash.
+func (r *RtmrExtend) ExtendDigest(hash []byte) error {
+	if (r.RtmrIndex != 2) && (r.RtmrIndex != 3) {
+		return fmt.Errorf("invalid rtmr index %d, userspace can only extend to rtmr2 or rtmr3", r.RtmrIndex)
```

Comment:
I don't think we should be making the decision to reject RTMR0 and RTMR1 extensions in the library, and instead allow the kernel to reject the writes. The error message could give a hint, "is your RTMR index extendable from userspace?".

Reply:
Done. Thanks for the suggestion.